src/rust/engine/serverset/src/lib.rs

```diff
@@ -220,65 +279,83 @@ impl<T: Clone + Send + Sync + 'static> Serverset<T> {
   pub fn next(&self) -> BoxFuture<(T, HealthReportToken), String> {
     let now = Instant::now();
     let mut inner = self.inner.lock();
-    let server_count = inner.connections.len();
+
+    let existing_connections = inner.connections.len();
+    if existing_connections < inner.connection_limit && inner.servers.len() > existing_connections {
+      // Find a healthy server without a connection, connect to it.
+      if let Some(ret) = inner.connect() {
+        inner.next_connection = inner.next_connection.wrapping_add(1);
+        return futures::future::ok(ret).to_boxed();
+      }
+    }
+
+    if !inner.connections.is_empty() {
+      let next_connection = inner.next_connection;
+      inner.next_connection = inner.next_connection.wrapping_add(1);
+      let connection = &inner.connections[next_connection % inner.connections.len()];
+      return futures::future::ok((
+        connection.connection.clone(),
+        HealthReportToken {
+          server_index: connection.server_index,
+        },
+      ))
+      .to_boxed();
+    }
 
     let mut earliest_future = None;
-    for _ in 0..server_count {
-      let i = inner.next % server_count;
-      inner.next = inner.next.wrapping_add(1);
-      let server = &inner.connections[i];
+    for _ in 0..inner.servers.len() {
+      let i = inner.next_server % inner.servers.len();
```

Comment:
This would read better if the wrapping iterator logic was implemented on a substruct of inner:
`inner.servers.iter().for_each(|server| ... )`

Reply:
As above